src/lib/comms/sol-mavlink.c

```diff
@@ -0,0 +1,786 @@
+        SOL_INF("Could not determine mode, we'll retry on next heartbeat");
+        return;
+    }
+
+    mavlink->mode = mode;
+    mavlink->target_system = msg->sysid;
+    mavlink->target_comp = msg->compid;
+    mavlink->type = type;
+
+    mavlink->base_mode = base_mode;
+    mavlink->custom_mode_enabled = base_mode & MAV_MODE_FLAG_CUSTOM_MODE_ENABLED;
+
+    mavlink->status |= SOL_MAVLINK_STATUS_INITIAL_SETUP;
+    sol_mavlink_request_home_position(mavlink);
+}
+
+static void
+sol_mavlink_heartbeat_handler(struct sol_mavlink *mavlink, mavlink_message_t *msg)
+{
+    uint8_t base_mode;
+    sol_mavlink_mode mode;
+
+    if (!(mavlink->status & SOL_MAVLINK_STATUS_INITIAL_SETUP)) {
+        sol_mavlink_initial_status(mavlink, msg);
+        return;
+    }
+
+    if (mavlink->target_system != msg->sysid ||
+        mavlink->target_comp != msg->sysid)
+        return;
+
+    mode = sol_mavlink_get_mode(mavlink->type, msg, &base_mode);
+    if (mavlink->mode != mode) {
+        mavlink->mode = mode;
+        CALL_HANDLER(mavlink, mode_changed);
+    }
+
+    if (mavlink->base_mode != base_mode) {
+        sol_mavlink_armed_transition(mavlink, base_mode);
+        mavlink->base_mode = base_mode;
+    }
+}
+
+static void
+sol_mavlink_position_handler(struct sol_mavlink *mavlink, mavlink_message_t *msg)
+{
+    int32_t lat, longi, alt;
+    struct sol_mavlink_position *pos = &mavlink->curr_position;
+
+    lat = mavlink_msg_gps_raw_int_get_lat(msg);
+    longi = mavlink_msg_gps_raw_int_get_lon(msg);
+    alt = mavlink_msg_gps_raw_int_get_alt(msg);
+
+    if (lat != pos->latitude || longi != pos->longitude || alt != pos->altitude) {
+        pos->latitude = lat;
+        pos->longitude = longi;
+        pos->altitude = alt;
+        mavlink->status |= SOL_MAVLINK_STATUS_GPS_SETUP;
+        CALL_HANDLER(mavlink, position_changed);
+    }
+}
+
+static void
+sol_mavlink_statustext_handler(mavlink_message_t *msg)
+{
+    char text[50];
+
+    mavlink_msg_statustext_get_text(msg, text);
+    SOL_DBG("%s", text);
+}
+
+static void
+sol_mavlink_home_position_handler(struct sol_mavlink *mavlink, mavlink_message_t *msg)
+{
+    struct sol_mavlink_position *pos = &mavlink->home_position;
+
+    pos->latitude = mavlink_msg_home_position_get_latitude(msg);
+    pos->longitude = mavlink_msg_home_position_get_longitude(msg);
+    pos->altitude = mavlink_msg_home_position_get_altitude(msg) * 1000;
+    pos->x = mavlink_msg_home_position_get_x(msg);
+    pos->y = mavlink_msg_home_position_get_y(msg);
+    pos->z = mavlink_msg_home_position_get_z(msg);
+
+    mavlink->status |= SOL_MAVLINK_STATUS_GPS_HOME_POS;
+}
+
+static bool
+sol_mavlink_fd_handler(void *data, int fd, uint32_t cond)
+{
+    struct sol_mavlink *mavlink = data;
+    mavlink_message_t msg;
+    mavlink_status_t status;
+    uint8_t buf[MAVLINK_MAX_PACKET_LEN];
+    int i, res;
+    sol_mavlink_status sol_status = mavlink->status;
+
+    bzero(buf, MAVLINK_MAX_PACKET_LEN);
```

Comment:
yep sure, the use of bzero and friends have been discussed by the other guys, I'm getting rid of it wherever appropriate.